### DISCUSSION TITLE
Can't connect to server through local or not local.

### DISCUSSION THREAD

**Discussion by CleavisX** [itzg/docker-minecraft-server#1133] [Q&A]
Hi!

I recently setup the itzg minecraft server docker on my QNAP NAS. I've gotten the server to startup and everything went smoothly. However, I cannot connect to the server using my public IP or localhost.

I'm not sure if this is a problem with my NAS settings or not.

**Comment by gundestrup** [comment]
You have to look at the docker settings for what the "external" ports are on your docker image.
I have selected fixet external port on the docker image.
Else the external docker port might change each time you restart the server.
Take a screenshot of your docker settings where port maping is visiable

**Comment by CleavisX** [comment]
> You have to look at the docker settings for what the "external" ports are on your docker image.
> I have selected fixet external port on the docker image.
> Else the external docker port might change each time you restart the server.
> Take a screenshot of your docker settings where port maping is visiable

First off, I'm using a QNAP NAS with container station to run the server.
![Capture](https://user-images.githubusercontent.com/78890653/107707777-42a5c400-6c77-11eb-9250-10c2ccc98d08.PNG)

Here is what shows when I check the docker settings. I am unsure of what the "host" section is and I assume the "container" section is the minecraft port, as its 25565. I am not sure how to set it to fixed external port.

**Comment by CleavisX** [comment]
Okay so I put the mode into bridge mode and gave it a static IP. I was able to connect locally using the IP I set it to (192.168.xx.xx) but how can I get it so people outside my network can join?

**Comment by itzg** [comment]
I would suggest deleting off the 2nd and 3rd port mappings and change the host port on the first to 25565:

![image](https://user-images.githubusercontent.com/988985/107718815-60d1eb00-6c9c-11eb-8272-d8ad91796151.png)

It's the first one you'll want to reference in your router's port forwarding and a fixed port value will be much easier to manage.

**Comment by gundestrup** [comment]
Also be carefull about an open server.
If you want to play with friends, then use whitelisting.

**Comment by JahCoby** [comment]
I have this issue currently where I can connect from any local machine but cannot connect through the public ip, or dns service, although I was able to before the docker method and simply running a server off of the windows machine. Currently the issue is happening with current docker version (everything as of 11/25/21) on currently updated Linux Mint xfce that I have running on a machine here. I have checked and double checked that the ports are forwarded from my router and did the port publishing when running the image, I have beat my head against every resource that was available but nothing suggested did anything, including trying the image again with -p for the 25565 ports, and attempting to check that docker would be listening on all hosts that was also suggested elsewhere, no dice on anything.

**Reply by itzg** [reply]
Can you provide more information with the compose file or docker run command you’re using? Also, what all things have you tried? Without know that it’s hard to suggest something that may not be a repeat.

**Reply by JahCoby** [reply]
EDIT It appears to have been resolved? No idea honestly what caused it to have the issue in the first place. But restarts/re-stalls of everything, and re-enabling all the firewalls/closing ports, to then reopen everything and try publishing ports in the container again whilst starting it back up seems to have resolved it. My guess is that the firewall or port forwarding simply had not properly updated the first few times? Kinda annoying after all that troubleshooting, but hey, looks fully functional lol. Appreciate you checking up on it with me while I was troubleshooting

<details>
  <summary>Previous comment</summary>
  This was the command I used to run the container {docker run -d -it -p 25565:25565 -e EULA=TRUE -e MEMORY=4G -e VERSION=SNAPSHOT -v /docks/servers/minecraft:/data --name mcserver itzg/minecraft-server}
I tried the docker ports command to check and see that the port was published (looks to be listening on 0.0.0.0 with port 25565), which appeared to be the case since I could connect to the server through the local IP of the host machine (192.168.x) but could not connect through my public IP nor the domain that forwards to my IP, tested a connection from outside my house and it failed. Docker is running on linux mint xfce. When the server was tested without docker, running "bare" on my windows computer it had no issues and outside players could connect with the ports forwarded on my router.

Troubleshooting wise, Firewall on the host machine is currently disabled completely, I have tried re-creating the container, checking the published ports, and I attempted a static ip with the "bridge" connection on linux mints internet settings, there was a guide saying that the image for dockers may need config to use the IPs but it was vague so didnt end up doing anything with that, and the server does work on the local network. I tried forwarding the ports on my router again, ensuring the host machine running docker specifically should have all ports open. I am currently looking to attempt running a server.jar bare on the machine instead of through docker to see if that changes anything.

edited to clarify some of the troubleshooting
</details>

**Reply by itzg** [reply]
Annoying it happened in the first place, but I'm glad it somehow resolved itself. I was also suspecting firewall config just slightly out of sync since everything else you did sounded right. For future reference a couple more troubleshooting things I use is:

Confirming port binding from OS perspective with:
```
sudo lsof -i:25565
```

To debug port forwarding itself, sometimes it is interesting to run a packet capture on the host machine with something like:
```
sudo tcpdump -i eno1 -v tcp port 25565
```
where you'd replace `eno1` with the physical ethernet port of your machine.

**Comment by Vincent-von-Schmidt** [comment]
I have the same problem, but mine doesn't fixed itself and I didn't find any solution. Although I reinstalled everything twice. Have anyone an idea why it's not working?

**Comment by mdjorov** [comment]
Hi. I also have this problem. On Windows machine I can't connect to the server from my hosting machine, from my network or from internet. 
The server is started as a service from yml file:
```
version: "3.8"

services:
  server:
    image: itzg/minecraft-server:latest
    tty: true
    stdin_open: true
    restart: unless-stopped
    ports:
      - 25565:25565
    environment:
      EULA: "TRUE"
      TZ: "Europe/Sofia"
      ENABLE_ROLLING_LOGS: "TRUE"
      LOG_TIMESTAMP: "TRUE"
      MOTD: "Welcome to Mishe's server"
      FORCE_GAMEMODE: "TRUE"
      MODE: "survival"
      PVP: "FALSE"
      LEVEL: "world"
      ONLINE_MODE: "FALSE"            
      OP_PERMISSION_LEVEL: 4
    volumes:
      - D:/Games/Minecraft Server/:/data
```
`docker service ls` shows the result from the image. I've disabled the firewall but this didn't help. 
![mc_service](https://github.com/itzg/docker-minecraft-server/assets/9763027/15e90fe2-e2d2-48ea-8215-c9135bb31918)

**Reply by itzg** [reply]
I know it's listed for bedrock, but see if this helps

https://github.com/PrismarineJS/bedrock-protocol/blob/master/docs/FAQ.md#cant-connect-to-localhost-win10-server-with-minecraft-win10-edition

**Comment by mdjorov** [comment]
Thank you for the answer. It was a different problem. I noticed that if I run the server as a container with docker run .... or with the yml file but with docker-compose I'm able to connect. It wasn't working only as service. I've tried postgres container as a service - again it wasn't working. So I've reinstalled docker on my windows PC and now it works great. Maybe something with the networks' management was messed up.
The server works, the kids are happy.
